src/services/middlewareServices/AuthenticationMiddlewareService.ts

```diff
@@ -10,63 +11,135 @@ import { AuthError } from "@errors/AuthError"
-  cookies: {
-    isomercms: string
-    isomercmsE2E?: string
-  }
+type UnverifiedSession =
+  | Partial<SessionData>
+  | { userInfo: Record<string, never> }
+  | { userInfo: SetOptional<SessionData["userInfo"], "email"> }
+
+export type VerifyAccessProps = UnverifiedSession & {
+  // NOTE: Either both properties are present on the cookie
+  // or none are present.
+  // We disallow having 1 or the other.
+  cookies: RequireAllOrNone<
+    {
+      isomercmsE2E: string
+      e2eUserType: string
+    },
+    "e2eUserType" | "isomercmsE2E"
+  >
   url: string
 }
 
+type TestUserTypes = "Email admin" | "Email collaborator" | "Github user"
+
+// NOTE: Precondition to use this function is that the user type is valid.
+const getUserType = (userType: string): TestUserTypes => {
+  if (userType === "Email admin") return "Email admin"
+  if (userType === "Email collaborator") return "Email collaborator"
+  if (userType === "Github user") return "Github user"
+  throw new Error(`Invalid user type: ${userType}`)
+}
+
+const extractE2eUserInfo = (userType: TestUserTypes): SessionDataProps => {
+  switch (userType) {
+    case "Email admin":
+      return {
+        isomerUserId: E2E_EMAIL_ADMIN_ISOMER_ID,
+        email: E2E_TEST_EMAIL,
+      }
+    case "Email collaborator":
+      return {
+        isomerUserId: E2E_EMAIL_COLLAB_ISOMER_ID,
+        email: E2E_TEST_EMAIL,
+      }
+    case "Github user":
+      return {
+        accessToken: E2E_TEST_GH_TOKEN,
+        githubId: E2E_TEST_USER,
+        isomerUserId: E2E_ISOMER_ID,
+        email: E2E_TEST_EMAIL,
+      }
+    default: {
+      const missingUserType: never = userType
+      throw new Error(`Missing user type: ${missingUserType}`)
+    }
+  }
+}
+
 export default class AuthenticationMiddlewareService {
-  verifyE2E({ cookies, url }: Omit<VerifyAccessProps, "userInfo">) {
-    const { isomercmsE2E } = cookies
+  private verifyE2E({
+    cookies,
+    url,
+  }: Omit<VerifyAccessProps, "userInfo">): TestUserTypes | false {
+    const { isomercmsE2E, e2eUserType } = cookies
     const urlTokens = url.split("/") // urls take the form "/v1/sites/<repo>/<path>""
 
-    if (!isomercmsE2E) return false
+    // NOTE: If the cookie is not set, this is an actual user.
+    if (!isomercmsE2E || !e2eUserType) return false
 
+    // NOTE: Cookie is set but wrong, implying someone is trying to figure out the secret
     if (isomercmsE2E !== E2E_TEST_SECRET) throw new AuthError("Bad credentials")
 
     if (urlTokens.length < 3) throw new BadRequestError("Invalid path")
 
+    const userType = getUserType(e2eUserType)
+
     // General access paths are allowed
-    if (GENERAL_ACCESS_PATHS.includes(url)) return true
+    if (GENERAL_ACCESS_PATHS.includes(url)) return userType
 
-    // Throw an error if accessing a repo other than e2e-test-repo
+    // Throw an error if accessing a repo other than
+    // the allowed repos for each respective user type
     const repo = urlTokens[3]
```

Comment:
Am confused here, where does this `3` come from ah

Reply:
from teh comment in the code, 

`// urls take the form "/v1/sites/<repo>/<path>"`

so the [3] refers to `repo` (we split by `/` so the output array is `['', 'v1', 'sites', '<repo>', '<path>']`)